### PR TITLE
add uploader compress writer

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
https://github.com/pingcap/dumpling/issues/7 

### What is changed and how it works?
Add uploader compress writer

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Related changes

 - Need to cherry-pick to the release branch

### Release Note

 - None

### CODE REVIEW

File: pkg/storage/writer.go [hunk 1]
```diff
@@ -54,6 +55,65 @@ func (u *uploaderWriter) Close(ctx context.Context) error {
...
 }
 
+type uploaderCompressWriter struct {
+	*uploaderWriter
+	bytesWritten int
+	cap          int
+	gzipWriter   *gzip.Writer
+}
+
+func (u *uploaderCompressWriter) Write(ctx context.Context, p []byte) (int, error) {
```
Comment (kennytm) [inline pkg/storage/writer.go:65]: 🤔 it is possible to make this function and `(*uploaderWriter).Write` share code? they look too similar.

Comment (lichunzhu) [reply pkg/storage/writer.go]: fixed

File: pkg/storage/writer.go [hunk 2]
```diff
@@ -2,11 +2,71 @@ package storage
 
 import (
 	"bytes"
+	"compress/gzip"
 	"context"
 )
 
+type interceptBuffer struct {
+	*bytes.Buffer
+	compressWriter *gzip.Writer
```
Comment (kennytm) [inline pkg/storage/writer.go:11]: make this an `interface { io.Writer; io.Closer; Flush() error }`, we may support other compression algorithms.

File: pkg/storage/writer.go [hunk 3]
```diff
@@ -2,11 +2,71 @@ package storage
...
+}
+
+func (b *interceptBuffer) Close() error {
+	if b.compressWriter == nil {
+		return nil
+	}
+	return b.compressWriter.Close()
+}
+
+func newInterceptBuffer(chunkSize int, compress bool) *interceptBuffer {
```
Comment (kennytm) [inline pkg/storage/writer.go:58]: use an enum instead of a bool

**Review by kennytm — COMMENTED** [review]
rest LGTM

**Review by kennytm — COMMENTED** [review]
LGTM but I wonder why do we want to support zlib :shrug: ("other compression algorithms" mostly means lz4 and zstd)

**Comment by lichunzhu** [comment]
> LGTM but I wonder why do we want to support zlib 🤷 ("other compression algorithms" mostly means lz4 and zstd)

Just because it's in compress package like `gzip`. Should I delete this?

**Comment by kennytm** [comment]
> Just because it's in compress package like `gzip`. Should I delete this?

gzip and zlib are equivalent except the container format, so yeah there's no point adding this. we're only interested in:

* gzip (legacy fast & low compression format)
* xz (somewhat-legacy slow & high compression format)
* lz4 (faster than gzip)
* zstd (faster than xz)

**Comment by kennytm** [comment]
@lichunzhu i mean we can keep `simpleCompressBuffer` just not add the `zlib` support 👀

**Comment by lichunzhu** [comment]
OK

**Comment by kennytm** [comment]
/run-integration-test

https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/br_ghpr_unit_and_integration_test/detail/br_ghpr_unit_and_integration_test/3609/pipeline

Hmm can't read the log at all.

```
Error: failed to validate checksum: [BR:Restore:ErrRestoreChecksumMismatch]restore checksum mismatch
```

**Comment by kennytm** [comment]
(retrigger CLA)

**Review by 3pointer — APPROVED** [review]
LGTM

**Pull request merged**